### ISSUE TITLE
[Feature Request] Want to specify and fixed the number of pages, copies, etc.

### ISSUE THREAD

**Issue by kiki-ja** [cleidigh/printing-tools-ng#139]
@cleidigh

Want to specify and fixed the number of pages, copies, etc.
This request was escalated by me picking up similar comments as follows.

Taken from the following.

**１．**[Feature Request/Question:: Set pages to custom #118](https://github.com/cleidigh/printing-tools-ng/issues/118#issue-1213246091)

@RRoenn wrote:
>does anyone know how to set the pages to custom by default so that you don't 
>have to switch them from print all pages everytime?

>I usually print only the first or first and second page.

**２．**[PrintingToolsNG v2.1.4 do not work with TB 102.2.0 when "print.prefer_system_dialog" is set to "true" #134](https://github.com/cleidigh/printing-tools-ng/issues/134#issuecomment-1227643847)

@ffesl wrote:
>    99% we only print the 1st page one-fold, so a shortcut to print only the 1st page of an email (with attachment list) would actually suffice.

**３．**[PrintingTools NG Reviews - addons.thunderbird.net](https://addons.thunderbird.net/en-US/thunderbird/addon/printingtools-ng/reviews/1170154/)

**dominik_99** wrote:
>Could you insert a function to print "Current Page Only" or "Print First Page Only" to avoid printing all pages in long emails without having to change the settings manually each time in the printing dialogue?

Please see the screenshots I have organized into screenshots.

![Specify the number of pages and other-0](https://user-images.githubusercontent.com/80428201/190418003-5969df47-5b67-4b26-98e7-00ae0b8bbc00.png)
![Specify the number of pages and other-2](https://user-images.githubusercontent.com/80428201/190418036-86c6f8e5-5b88-473f-ae3e-ec34a51367d1.png)

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG-2.1.5 **pdft6**
Thunderbird 91.13.1, 102.5.1 64bit
Windows 10 Pro 22H2 64bit
CubePDF Virtual Printer

#152 mentions.
Here is a report on the test version.

- Regarding the 'Page Options' section
I think good with 4（Copies, Page Ranges, Margins, Headers & Footers）.
However, those who prefer automatic printing (e.g. by FiltaQuilla) may find them insufficient.

- About UI
It is now compact and easy to see.
I prefer compact.

**Operation Test Result:**
1. Copies
TB 91.13.1 - No more than 2 copies will be printed.
If the Page Ranges of PTNG is 'Custom(e.g. by 1)' + 'Copies(e.g. by 2 higher)', Copies in Print preview will always be '1'.
TB 102.5.1 - Same as above.
*Please see screenshots.
　![Copies](https://user-images.githubusercontent.com/80428201/206825098-c94d9e31-453b-4b63-9ac9-e40d06acad24.png)

2. Pages (Page Ranges)
TB 91.13.1 - Prints correctly.
TB 102.5.1 - Same as above.

3. Margins
TB 91.13.1 - prints correctly.
TB 102.5.1 - Same as above.

4. Headers & Footers
TB 91.13.1 - Prints correctly.
TB 102.5.1 - Same as above.

5. 'Empty' in Page Range
I think it is very good that 'Empty page range set to All' can be set to Default 'All'.
*Please see screenshots.
　![Empty page range set to All-0](https://user-images.githubusercontent.com/80428201/206825118-162f4ac2-76d7-41b3-9230-659e31eb6535.png)

**Wish to improve point:**
- Notation of Page Ranges setting values
I would like PTNG and Print preview dialog to have the same notation.
*Please see screenshots.
　![Hope Same Page Ranges](https://user-images.githubusercontent.com/80428201/206825455-042410e5-3fc1-4426-9b69-dd1a8348ffa6.png)

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG-2.1.5 **pdft7**
Thunderbird 91.13.1, 102.6.1(RC1) 64bit
Windows 10 Pro 22H2 64bit
CubePDF Virtual Printer

>Wish to improve point:
>- Notation of Page Ranges setting values
>　I would like PTNG and Print preview dialog to have the same notation.

I have confirmed that the same notation is now used.
Thank you very much.

![Page Ranges 1,2,3 - t7](https://user-images.githubusercontent.com/80428201/208680567-8797014e-13c5-4183-818b-26d5fe513af6.png)

![Page Ranges 1 - t7](https://user-images.githubusercontent.com/80428201/208680594-b83e1dbb-75e5-4fd9-82cf-6b3fd3fe0ccc.png)

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG-2.2.0 b5
Thunderbird 91.13.1, 102.6.1 64bit
Windows 10 Pro 22H2 64bit
CubePDF Virtual Printer

>Operation Test Result:
>
>    1. Copies
>    TB 91.13.1 - Can't print more than 2 copies.
>    If the Page Ranges of PTNG is 'Custom(e.g. by 1)' + 'Copies(e.g. by 2 higher)', Copies in Print preview will always be '1'.
>    TB 102.5.1 - Same as above.

Once the above issue are resolved, this issue is complete.

@kiki-ja

**Comment by cleidigh** [comment]
@kiki-ja 
This is related to a PDF driver handling question I have thought about before.  The issue is that I do not think PDF output ever does more than one copy output. This makes sense because why do you need more than one digital copy. Also you would have to name each  output.  In the TB code I found it hides the copies entry for known PDF drivers, CubePDF is not one of them.  So I also confirmed that CubePDF will only do one copy regardless of what is set in TB preview. I tested with PTNG disabled.

So I propose to disable the copies field for PDF printers and fix it at one.
Thoughts?
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh
To All

>The issue is that I do not think PDF output ever does more than one copy output. This makes sense because why do you need more than one digital copy. Also you would have to name each output.

I didn't know that was the specification.
I may need to add a note to PTNG Help about the differences with the driver of the printer device that prints on paper.

>So I also confirmed that CubePDF will only do one copy regardless of what is set in TB preview. I tested with PTNG disabled.

I also confirmed this.
Is this a common specification for Virtual Printer such as CubePDF?
I would like to see test reports from various printer users who print on paper.

>So I propose to disable the copies field for PDF printers and fix it at one.

I agree.

[CubePDF]
CubePDF: Virtual PDF printer
https://www.cube-soft.jp/en/cubepdf/index.php

cube-soft/cube.pdf: CubePDF, CubePDF Utility, CubePDF Page, and CubePDF SDK for .NET
https://github.com/cube-soft/cube.pdf
cube.pdf/Readme.md at master · cube-soft/cube.pdf
https://github.com/cube-soft/cube.pdf/blob/master/Readme.md

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh
To All.

PTNG-2.2.0 b7
Thunderbird 91.13.1, 102.7.0 RC1 64bit
Windows 10 Pro 22H2 64bit
CubePDF Virtual Printer 3.0.1

Pages setup (PTNG) is different from print preview (TB), is this correct behavior?

Please see screenshots.
![220b7-Pages option](https://user-images.githubusercontent.com/80428201/212927720-31ace550-f2a4-44d0-9e16-e2b301d31bca.png)

@kiki-ja

**Comment by cleidigh** [comment]
@kiki-ja 
As I said in my email this is a big bug. I have not seen this so hopefully if you capture the log we can learn something.  When the preview opens I have a listener that takes our settings and overwrites the preview settings.
Either the listener did not fire or the listener failed to change the settings.
@cleidigh

**Comment by RRoenn** [comment]
Hello @all
I did't have this bug with TB 91.13.1 and PTNG 2.2.0-b6
Some quick test looks very good

 ( uh, I got an old beta? from v2.2.0 Release Status #157 )

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-**b7**
Thunderbird 91.13.1, 102.7.0 RC1 64bit
Windows 10 Pro 22H2 64bit
CubePDF Virtual Printer 3.0.1

Attached is the console log.
I have checked for conflicts with other add-ons and there are none.

[Thunderbird 91.13.1 - PTNG 2.2.0-b7]
[console-export-2023-1-18_8-32-26.txt](https://github.com/cleidigh/printing-tools-ng/files/10446464/console-export-2023-1-18_8-32-26.txt)
![console 2023-1-18 TB91 13 1](https://user-images.githubusercontent.com/80428201/213177762-688068c2-4f60-4ca5-b6ea-fb70977b08f1.png)

[Thunderbird 102.7.0 RC! - PTNG 2.2.0-b7]
[console-export-2023-1-18_8-45-5.txt](https://github.com/cleidigh/printing-tools-ng/files/10446467/console-export-2023-1-18_8-45-5.txt)
![console 2023-1-18 TB102 7 0](https://user-images.githubusercontent.com/80428201/213177802-7bc00023-89bb-489d-9440-afe810312bd1.png)

@RRoenn 

As per your test report.
PTNG 2.2.0-b6 is normal and the problem is happening in PTNG 2.2.0-b7.

@kiki-ja

**Comment by cleidigh** [comment]
@RRoenn 
@kiki-ja 
Thank you both. I don't understand why I cannot reproduce.
Does this happen 100% of the time?
@cleidigh

**Comment by cleidigh** [comment]
@RRoenn 
@kiki-ja 
This is a scope issue. I am confused because I cannot understand why you would see a difference in scope then myself.
I made a small change how printingtools is defined.
b8 
- #157 

@cleidigh

**Comment by cleidigh** [comment]
This is my screenshot, you can see the width is not too wide.
Something different happens with @kiki-ja 

<img width="960" alt="2023-01-18-15-31-53 " src="https://user-images.githubusercontent.com/25272315/213289268-416b0ae5-9b7f-459e-8b17-5cb3016ce94a.png">

@cleidigh

**Comment by cleidigh** [comment]
@RRoenn 
@kiki-ja 
I was finally able to reproduce the failure with b7 and fix with b8.
I think you will see the same.
If so we just need to figure out the ui width issue and then final steps.
@cleidigh

**Comment by RRoenn** [comment]
![grafik](https://user-images.githubusercontent.com/77779094/213302061-9fea4ae9-1b45-455e-ae98-93e1b4b815d6.png)

![grafik](https://user-images.githubusercontent.com/77779094/213302136-3c61693f-360c-4551-af62-1ebefc2624b4.png)

![grafik](https://user-images.githubusercontent.com/77779094/213302211-1d654fdc-916b-4280-adb6-5fd7634cb8d3.png)

![grafik](https://user-images.githubusercontent.com/77779094/213302259-8cb634da-9bc0-492d-ae88-f1ab81de0f0f.png)


If you need more shoots, let me know.

Second tab, the right border cut some text.

**Comment by cleidigh** [comment]
@RRoenn 
Thanks those are perfect.
They show just about the same as my us rendering.
As for b8 that is good, I expected that.
Hey Dark mode looks pretty sweet  huh 👍 
lol my eye gaze keyboard does not have the win key!
@cleidigh

**Comment by cleidigh** [comment]
@kiki-ja 
I have an idea why your ui is stretched.
I could not test, but if you want to try here is what I think.
The input for the pdf output directory is making your width too wide. I think this may be because I use the imprecise size attribute for 58 characters. If you remove that and ad a fixed pixel width I think we will be close to parity:
style="width: 395px"

So remove size attribute and replace with the style attribute above.
@cleidigh

**Comment by RRoenn** [comment]
> @RRoenn  my eye gaze keyboard does not have the win key! @cleidigh

you can try Snip & Sketch App
https://apps.microsoft.com/store/detail/ausschneiden-und-skizzieren/9MZ95KL8MR0L

or https://www.thewindowsclub.com/snipping-tool-capture-screenshots-windows-8-tips

**Comment by kiki-ja** [comment]
@cleidigh

In 2.2.0-b7 it was 100% reproduced.
It worked fine in 2.2.0-b8 and resolved.
Thank you very much.

![220b8-Pages-OK](https://user-images.githubusercontent.com/80428201/213459645-05ab8410-c475-4d95-8ab4-fda072b88f79.png)

@kiki-ja

**Comment by cleidigh** [comment]
@kiki-ja 
I am online for just a minute.
That's good, I was concerned. It was an artifact from debugger sequence.
Hopefully my thoughts on the ui are correct also so we can finish ui tweaks.
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh

I changed size="58" in the input area to style="width: 395px" as suggested.
Other areas were similarly changed to style attribute with the same ratio of size="58" / style="width: 395px".

The result is shown in the screenshot.

![change size en-US](https://user-images.githubusercontent.com/80428201/213477928-3fb01fa6-900a-49c9-b412-bc28122c87fb.png)

![change size ja](https://user-images.githubusercontent.com/80428201/213477984-9c29469d-9398-4cf5-b15c-cbc72bcf817f.png)

I do not know what causes the UI to spread horizontally.
I have not changed the DPI and scale of my OS (Windows 10 Pro).
However, the English and Japanese versions of the same Thunderbird are slightly different in size.
The same problem may exist in the CJK area.
Also, it may also be due to the difference in units (meters and inches) on the OS.

I don't know about 'the imprecise size attribute', but maybe 'the style attribute' is better.
Or maybe it is better to define and specify it in CSS.

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-b8 and b9
Thunderbird 91.13.1, 102.7.0 64bit
Windows 10 Pro 22H2 64bit

Direct input of margins [.] is not included
When I try to type 10.00, I cannot type [.]. 
The result is 10000.00.

![Margins - input value](https://user-images.githubusercontent.com/80428201/213694179-0643c703-033b-428f-bcdf-c84c0c5aec2e.png)

Is this a specification?

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-b9
Thunderbird 91.13.1, 102.7.0 64bit
Windows 10 Pro 22H2 64bit

>We can determine the best pixel width for the input elements for ja independent of en or other locales.

Thanks for the consideration for the ja locale.
The windows are now the same size.

>I have the background color just as a temporary visual indicator the rule gets recognized.

I feel that the background color of the release version should be the same as the en-US version or none.

![220b9-en-US-tab4](https://user-images.githubusercontent.com/80428201/213730326-ff10764a-3a4d-4f6e-b361-cd88c95fed8f.png)

![220b9-ja-tab4](https://user-images.githubusercontent.com/80428201/213730340-2dfaf759-06d2-4b1f-8444-d03726b37827.png)

-------------

I think the UI should be discussed in another Issue.
What do you think?

@kiki-ja

**Comment by cleidigh** [comment]
@kiki-ja 
Agreed, created 
- #159 
See comments there.

I confirmed the decimal issue, will fix next beta.
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh

I have confirmed that the decimal issue has been fixed and resolved in PTNG 2.2.0-b11.
Thank you very much.

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-**b19**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Virtual printer: CubePDF, **Mozilla PDF**

1. Set page to 1 in PTNG.
2. Open print preview and set page = 1, but preview still shows 1/5.
3. When printed, it is 5 pages.

**[PTNG pages =1]**
![220b19-tab5-pages](https://user-images.githubusercontent.com/80428201/217708431-ba9eda76-3e94-4d5d-ac25-f22b50f31a2c.png)

**[Print preview - pages 1/5]**
![220b19-print-preview](https://user-images.githubusercontent.com/80428201/217708443-081ea9de-851e-4950-b76d-72c04ab1c08e.png)

Have the specifications changed since the previous beta version?

@kiki-ja

**Comment by cleidigh** [comment]
ugg I can't keep up!

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-**b17**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Printer: Save to PDF / Mozilla_Save_to_PDF

In 2.2.0-**b17** the preview page display are normal working.
Print output it are normal working too.
It looks like the margin setting changes are working too.
It also appears to correctly recognize the difference between inch and mm (paper_size_unit).

**[PTNG set pages custom= 1]**
![220b17-ptng-pages](https://user-images.githubusercontent.com/80428201/218295602-2679c37a-8220-4230-be00-f1c6510dc70f.png)

**[Print preview - Pages custom=1 and display 1 of 1]**
![220b17-printpreview-pages](https://user-images.githubusercontent.com/80428201/218296279-2bdab00b-887a-4628-832a-915a8a21818d.png)

**[print_paper_size_unit]**
![paper_size_unit](https://user-images.githubusercontent.com/80428201/218295889-1310bfac-2c5e-4c3a-bf99-e0023b1d40d9.png)


If the pages, margins, and unit(inch and mm) problems are caused by the printer driver, I feel there is nothing can do on the PTNG side.

@kiki-ja

**Comment by RRoenn** [comment]
@kikiThe problem is not with the printer driver. Please try to change margins with mm.All your printer should be in mm expect Mozilla printer driver unit is 0

**Comment by kiki-ja** [comment]
@RRoenn
@cleidigh

In my case, Save to PDF (Mozilla_Save_to_PDF) will always be inch because unit is 0.
Since the default is inch, I don't bother to change it to mm.

>please check after modify margins about:config() printer_margin and about:config() ptng settings and also print preview.

Both values will be the same, as set.

ーーーーーー

@kiki-ja wrote.
>If the pages, margins, and unit(inch and mm) problems are caused by the printer driver, 

@RRoenn wrote.
>The problem is not with the printer driver. 

I am not necessarily assuming that the printer driver is the cause.
It is just that different printers may behave differently in inch and mm.

This can be seen by testing without PTNG.
In the case of Save to PDF (Mozilla_Save_to_PDF), inch remembers the margin values when printing, but mm resets them every time.

This leads me to believe that it could be a printer driver issue.

@kiki-ja

**Comment by RRoenn** [comment]
@kiki-ja 
please try this (with or wihout PTNG)

1. go to print preview
2. select a printer with **mm** (not Mozilla printer!)
3. select papersize A4
4. change all margins to 50 mm
5. you can print or cancel no matter what
6. check and report the values in about:config() **print.printer_YOUR_Printer.print_margin_*** 

thank you

**Comment by tidebre67** [comment]
@RRoenn 
@cleidigh 
shows the fllwg, guess inch:
print.printer_PDF24.print_margin_bottom | 1.96875 |

**Comment by kiki-ja** [comment]
@RRoenn
@cleidigh

PTNG 2.2.0-**b17**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Printer: CubePDF (mm)

>please try this (with or wihout PTNG)

Results of the trial:

1. Without PTNG
print.printer_CubePDF.print_margin_bottom	1.96875
print.printer_CubePDF.print_margin_left	1.96875
print.printer_CubePDF.print_margin_rightt	1.96875
print.printer_CubePDF.print_margin_top	1.96875

2. With PTNG 2.2.0-b17
print.printer_CubePDF.print_margin_bottom	1.96875
print.printer_CubePDF.print_margin_left	1.96875
print.printer_CubePDF.print_margin_right	1.96875
print.printer_CubePDF.print_margin_top	1.96875

extensions.printingtoolsng.printer.CubePDF	{"numCopies":1,"pageRanges":[1,1],"marginTop":1.96875,"marginBottom":1.96875,"marginLeft":1.96875,"marginRight":1.96875,"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}

[about:config]
![CubePDF Margins](https://user-images.githubusercontent.com/80428201/218316474-8f0a3fc9-8157-4e64-ac09-165d47ec9ff7.png)

![PTNG margins](https://user-images.githubusercontent.com/80428201/218316485-05ae0758-7711-4e48-a616-62f30c8e373a.png)

[PTNG tab5 - set margins ]
![PTNG tab5-margins](https://user-images.githubusercontent.com/80428201/218316493-5b5ce165-a5b6-466a-8400-0dbaa59a494b.png)

@kiki-ja

**Comment by tidebre67** [comment]
> @RRoenn @cleidigh shows the fllwg, guess inch: print.printer_PDF24.print_margin_bottom | 1.96875 |

@RRoenn 
@cleidigh 
after restart config still 1,96875, but print preview back to default.

**Comment by tidebre67** [comment]
> > @RRoenn @cleidigh shows the fllwg, guess inch: print.printer_PDF24.print_margin_bottom | 1.96875 |
> 
> @RRoenn @cleidigh after restart config still 1,96875, but print preview back to default.

after new restart changed in print preview the default figures manually with the default figures (12.5) and the config shows the default margins again.

**Comment by kiki-ja** [comment]
@RRoenn
@tidebre67
@cleidigh

PTNG 2.2.0-**b17**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Printer: CubePDF (mm)

Results of the trial:
After restarting Thunderbird, the configuration values (about:config, Print preview and PTNG set) are unchanged and the same.

Add screenshot.

[Print preview]
![220b17-print preview-1](https://user-images.githubusercontent.com/80428201/218318791-0030e71e-ac23-43bf-8b96-98a020c4329d.png)

![220b17-print preview-2](https://user-images.githubusercontent.com/80428201/218318803-e95f0853-e5c7-49b8-871c-eb8c40af7247.png)


@kiki-ja

**Comment by tidebre67** [comment]
@kiki-ja 
@cleidigh 
@RRoenn 
that's really strange pls see my comment above. But i am on b21.

**Comment by kiki-ja** [comment]
To All.

Please that you write the environmental conditions according to PTNG help.
e.g PTNG 2.2.0-b17, Thunderbird 102.7.2, Windows 10 22H2, and etc.

@kiki-ja

**Comment by RRoenn** [comment]
@tidebre67 @kiki-ja thank you

@cleidigh 
As expected, tb stores the margins in inches, regardless of whether unit is set to 1 or 0.
@tidebre67 The fact that the margins are reset is due to the wrong setting in PTNG. I hope for a bug fix soon.
Mixed inches and mm

**Comment by kiki-ja** [comment]
@tidebre67 

I am taking up and mentioning PTNG 2.2.0-**b17**.
Not b-21.

@kiki-ja

**Comment by tidebre67** [comment]
@cleidigh 
@RRoenn 
@kiki-ja 

just found the fllwg:
I changed mm margins (50mm) in the ptng print settings and the config shows the correct 50mm figures.
But when i change the margin figures  50 in the config back to default mm (12.7) the margins in the ptng print settings
remain 50mm.
The print preview margins never changed and remain  at 12.7.

**Comment by RRoenn** [comment]
> The print preview margins never changed and remain at 12.7.

PTNG saves mm, used at inch in print preview and they are to big -> tb set to default

**Comment by tidebre67** [comment]
@cleidigh 
@RRoenn 
@kiki-ja 
Christopher, when i read all above correct and taking all into account there is not only 1 bug but at least 3.

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-**b17**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Printer: PDF24 Creator version 11.10.2 (https://www.pdf24.org/)

Here is a test report on the PDF24 Creator printer.
Margins was set to inch.
The results are as follows.
No problem, normal, no reset.

[unit value]
print.printer_PDF24.print_paper_size_unit	0 (inch=default)

[margin values]
print.printer_PDF24.print_margin_bottom	0.300000011920929
print.printer_PDF24.print_margin_left	0.300000011920929
print.printer_PDF24.print_margin_right	0.300000011920929
print.printer_PDF24.print_margin_top	0.300000011920929

[PTNG set value]
extensions.printingtoolsng.printer.PDF24	{"numCopies":1,"pageRanges":[1,1],"marginTop":0.30000001192092896,"marginBottom":0.30000001192092896,"marginLeft":0.30000001192092896,"marginRight":0.30000001192092896,"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}

![220b17-pdf24-margins](https://user-images.githubusercontent.com/80428201/218441150-ac04af41-3b24-4be7-9cd5-89ead5c5679d.png)

![pdf24-printpreview](https://user-images.githubusercontent.com/80428201/218441177-df355ab6-dc39-46b0-b792-bc18869d0ff3.png)

@kiki-ja

**Comment by RRoenn** [comment]
@kiki-ja 
there is no problem with inch. But a prove, that PNG is working correctly. The problem is with mm printers, because of the wrong saving of the units. PTNG save mm but have to save inches.

**Comment by tidebre67** [comment]
@RRoenn 

confirm, if print_paper_size_unit 0 used all works perfect.
MM printer are the problem as mentioned by RRoenn.

**Comment by kiki-ja** [comment]
@RRoenn
@tidebre67
@cleidigh

I don't care whether the margin setting values are saved in inches or mm.
As long as they are displayed correctly in the PTNG options and print preview dialogs and print correctly, I am fine.

Fortunately, PTNG 2.2.0-**b17** displays and prints correctly in my **CubePDF(mm)** Mozilla Save to PDF(inch) and PDF24 Creator(inch).
All units (inch and mm) remain default and I have not changed them.

@kiki-ja

**Comment by kiki-ja** [comment]
@RRoenn
@tidebre67

My case.I will repeat it again.

PTNG 2.2.0-**b17**
Thunderbird 102.7.2 64bit
Windows 10 Pro 22H2 64bit
Printer: CubePDF(**mm**)

[CubePDF printer]
- print_paper_size_unit = 1 (**mm**)
- paper size = A4
- margins = all ５０**mm**
- print.printer_CubePDF.print_margin_* = 1.96875(inch)
- extensions.printingtoolsng.printer.CubePDF = margin* 1.96875(inch)

[print preview and print]
- margins = all 50(**mm**)

**As introduced below.**
https://github.com/cleidigh/printing-tools-ng/issues/139#issuecomment-1426953351
https://github.com/cleidigh/printing-tools-ng/issues/139#issuecomment-1427046024
https://github.com/cleidigh/printing-tools-ng/issues/139#issuecomment-1427055296

[Add infomation]
I tested PDF24 Creator with print_paper_size_unit = 1 (**mm**) and got the same results as above.

Have you two tested with PTNG 2.2.0-**b17**?

@kiki-ja

**Comment by kiki-ja** [comment]
@RRoenn
@tidebre67
@cleidigh

What I am trying to tell you is that PTNG 2.2.0-**b17** saves margin value as inch, but in print preview and print it is **correct mm**.

I do not know why this is so.
Maybe it is caused by differences in Locale, OS locale and version, yard-pound system vs. metric system, printer driver, Thunderbird locale, Language pack, other differences.

@kiki-ja

**Comment by RRoenn** [comment]
Yes, I will test -b17 again tonight. I'm just afraid we lost Chrisopher...

**Comment by cleidigh** [comment]
@cleidigh 
Gentleman, sorry for being awol, bad day yesterday and still a little slow.
My advice for the day is don't get ALS, it sucks %-{

I really need a vote on continuing to get margins working reliably. While I am a bit frustrated, I am more concerned about the burden on you three for testing. If all of you wish to continue then we will. 

Some comments : 
- The copies and page ranges, headers and footers appear to be working 
- Both copies and page ranges had nasty timing issues as to when they needed to be set after the preview was shown
- I fixed copies in b19 then @RRoenn reported problems with page ranges 
- I believe I have fixed that, but requires more testing in next b22
- b21 changed from b17 in that I stored margin values as their matching units, b21 always stored inches
- presentation in b17 ptng was driven by the paper units found in the printer preferences (one bug was displaying inches for the Mozilla pdf driver in your metric locales)
- PLEASE LET ME KNOW IF ABOVE IS INCORRECT 
- There are several edge conditions I know we have in "time zero", meaning before any printer is configured 
- There is also the scenario where the user changes paper types with different units in preview or the margins
- I have intended that the ptng margins would always take precedence over one time changes in preview - THIS MAY BE A WRONG CHOICE 
- Storing the margins in ptng is probably not necessary if we always use whatever is set in preview
- I think this would simplify the code and behavior 
- In this mode think of ptng as an alternative way to set the margins from preview, however, ptng would not change margins different from preview 
- If anyone is unclear on this ask me
- At this point the question is do we miss any added functionality by using this different behavior?

Christopher

**Comment by cleidigh** [comment]
@kiki-ja 
Did I understand that cube24 has some papers stored as inches, but displayed in mm for locale?
If this is the case I have to find a Javascript way to determine the number formatting for the locale, I think that should not be too hard.
This is definitely a current issue 
@cleidigh

**Comment by cleidigh** [comment]
@RRoenn 
If we are going forward I would not do much with b17 as the methods will change. Most importantly just need to understand that mm locales always display mm in preview regardless of paper units.
@cleidigh

**Comment by cleidigh** [comment]
@RRoenn 
@kiki-ja 
@tidebre67 
Don't do anything with b17, if everyone wants to continue, it will be too different from b22, take a break.
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh

>Did I understand that cube24 has some papers stored as inches, but displayed in mm for locale?

Yes, I understand that it is internally converted from inch to mm to match the locale and displayed.

>If this is the case I have to find a Javascript way to determine the number formatting for the locale, I think that should not be too hard.

I don't know Javascript, but I feel that it would be good to find a BETTER strategy.

ーーーー

I think the current beta has reached the point where Copies, Page ranges, Headers and Footers are working correctly.
I think the remaining issue is Margins.

Fortunately, in my case, it is working correctly on b17, but on the other hand, there seems to be a problem that it does not work correctly on b21, so I hope it will be improved in the future.

@kiki-ja

**Comment by RRoenn** [comment]
@tidebre67 
> But when i change the margin figures 50 in the config back to default mm (12.7) the margins in the ptng print settings
> remain 50mm.

Question, what do you mean by config? Print preview or about:config() settings?
It is not intended that the print preview changes the settings in PTNG, the values are intended to be fixed on purpose.

**Comment by RRoenn** [comment]
@cleidigh 
I am happy that you are back again. Take your time.

**Comment by tidebre67** [comment]
> @tidebre67
> 
**> > But when i change the margin figures 50 in the config back to default mm (12.7) the margins in the ptng print settings
> > remain 50mm.**
> 
> Question, what do you mean by config? Print preview or about:config() settings? It is not intended that the print preview changes the settings in PTNG, the values are intended to be fixed on purpose.
> 
> _edit_ PTNG settings are stored in **extensions.printingtoolsng.printer.YOUR_PRINTER** print preview settings are stored in **print.YOUR_PRINTER.print_margin_***
> 
> e.g. [margin values] print.printer_PDF24.print_margin_bottom 0.300000011920929 print.printer_PDF24.print_margin_left 0.300000011920929 print.printer_PDF24.print_margin_right 0.300000011920929 print.printer_PDF24.print_margin_top 0.300000011920929
> 
> [PTNG set value] extensions.printingtoolsng.printer.PDF24 {"numCopies":1,"pageRanges":[1,1],"marginTop":0.30000001192092896,"marginBottom":0.30000001192092896,"marginLeft":0.30000001192092896,"marginRight":0.30000001192092896,"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}

>**> > But when i change the margin figures 50 in the config back to default mm (12.7) the margins in the ptng print settings
> > remain 50mm.**<

@RRoenn 
config = config editor

i meanwhile stopped testing as Christopher mentioned to take a break. I'll wait till b22 is out.
Currently i am totally confused of all informantion.

**Comment by RRoenn** [comment]
@cleidigh 

> * I have intended that the ptng margins would always take precedence over one time changes in preview - THIS MAY BE A WRONG CHOICE

I think this was the desired function - and not to have an equivalent to the print preview

**Comment by RRoenn** [comment]
@tidebre67 

Yes, we will do a break and waiting for b22

> Currently i am totally confused of all information.

Maybe I can help you? Gern auch in deutsch ...

**Comment by tidebre67** [comment]
@RRoenn 

>Maybe I can help you? Gern auch in deutsch ...<
see here for contact:
https://github.com/tidebre67

**Comment by RRoenn** [comment]
@tidebre67 

> But when i change the margin figures 50 in the config back to default mm (12.7) the margins in the ptng print settings
> remain 50mm.

Just tried and all worked like expected.
you have to change about:config **extensions.printingtoolsng.printer.YOUR_PRINTER** 
e.g. 
```
marginLeft":0.30000001192092896 to
marginLeft":1,
```

Be careful when editing, there is a colon (:) before the number.
In the ptng print settings you have to select also YOUR_PRINTER to check the changes. I had overlooked that at first glance.

@RRoenn

**Comment by cleidigh** [comment]
@tidebre67 
@RRoenn 
@kiki-ja 
Ok as people agreed, I am moving forward with b22, I actually made a lot of  progress yesterday.

Let me summarize where I am : 
- I  determine localeUnits for presentation by the following logic
- If locale == en-US use inches else use mm
- I checked, only us nitwits in US along with Myanmar use imperial units
- This resolved a key bug disconnecting paper units from presentation units.
- I do conversion if localeUnits != paper units 
- KEY CHANGE - I no longer store margins in ptng printer settings 
- Now this means preview margins changes will be persistent 
- This removes what @RRoenn mentioned above, using ptng as sticky defaults
- let me get b22 out this way and get the basic problems solved, then look at making margins sticky again 
- b22 should address the page ranges issue @RRoenn found 

@cleidigh

**Comment by cleidigh** [comment]
@kiki-ja 
@tidebre67 
@RRoenn 
ok b22

- multiple changes to preferences handling 
- fix for pageRanges not always being set or set and not reflected
- enable either period or comma for decimal in margins
- use localeUnits to control presentation 
- don't save margins in ptng 

Things b22 does not handle yet
- tb start where no printer is set
- changes in preview overwrie ptng settings (replaced)

I wanted to do more,but didn't want to miss testing tonight 

BEFORE STARTING 
delete all ptng printer preferences 

Focus on proving the margins are handled correctly.
If we are good I will make them sticky and reset before print to ptng settings 

Thanks for your perseverance ! 
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh

Test Result of PTNG 2.2.0-**b23**

Thunderbird 102.8.0 64bit
Windows 10 Pro 22H2 64bit

Skipped the testing of PTNG 2.2.0-b22.
In my environment PTNG 2.2.0-b23 displayed Copies, Page ranges and Marigins correctly.

**[paper_size_unit / about:config]**
print.printer_CubePDF.print_paper_size_unit	1(mm)
print.printer_Mozilla_Save_to_PDF.print_paper_size_unit	0(inch)
print.printer_PDF24.print_paper_size_unit	1(mm)

**[PTNG Set vlaues - Copies=1, Page ranges=2,3 and Margins=10mm / about:config]**
extensions.printingtoolsng.printer.CubePDF	{"numCopies":1,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}	
extensions.printingtoolsng.printer.Mozilla_Save_to_PDF	{"numCopies":1,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}	
extensions.printingtoolsng.printer.PDF24	{"numCopies":1,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}
extensions.printingtoolsng.printer.MS_Publisher_Color_Printer	{"numCopies":2,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}	
extensions.printingtoolsng.printer.Microsoft_Print_to_PDF	{"numCopies":1,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}	
extensions.printingtoolsng.printer.Microsoft_XPS_Document_Writer	{"numCopies":2,"pageRanges":[2,2,3,3],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}

**[Margins / about:config]**
print.printer_CubePDF.print_margin_bottom	0.393750011920929
print.printer_CubePDF.print_margin_left	0.393750011920929
print.printer_CubePDF.print_margin_right	0.393750011920929
print.printer_CubePDF.print_margin_top	0.393750011920929

print.printer_Mozilla_Save_to_PDF.print_margin_bottom	0.393750011920929
print.printer_Mozilla_Save_to_PDF.print_margin_left	0.393750011920929
print.printer_Mozilla_Save_to_PDF.print_margin_right	0.393750011920929
print.printer_Mozilla_Save_to_PDF.print_margin_top	0.393750011920929

print.printer_PDF24.print_margin_bottom	0.393750011920929
print.printer_PDF24.print_margin_left	0.393750011920929
print.printer_PDF24.print_margin_right	0.393750011920929
print.printer_PDF24.print_margin_top	0.393750011920929

**Note:**
Copies set in PTNG were not reflected only for MS_Publisher_Color_Printer.
Also, paper_size_unit item was not.

@kiki-ja

**Comment by RRoenn** [comment]
@kiki-ja 
what do you mean with that note?

> Also, paper_size_unit item was not.

**Comment by kiki-ja** [comment]
@RRoenn

>is this related to #161 ?

I do not know.
It may be related, it may be irrelevant.

>what do you mean with that note?

It does not mean anything.
I simply wrote the results of my check.
I do not know the cause or reason.

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-b24
Thunderbird 102.8.0 64bit
Windows 10 Pro 22H2 64bit

This is a test report for each printers.

**[Mozilla Save to PDF]**
extensions.printingtoolsng.printer.Mozilla_Save_to_PDF = {"numCopies":1,"pageRanges":[1,1],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}
print.printer_Mozilla_Save_to_PDF.print_margin_* = 0.393750011920929 (inch)
print.printer_Mozilla_Save_to_PDF.print_paper_size_unit = 0 (inch)
print.printer_Mozilla_Save_to_PDF.print_paper_id = iso_a4 (A4)
![220b24-PTNG-Mozilla Save to PDF-margins](https://user-images.githubusercontent.com/80428201/220107943-cd71391a-4439-4806-aba9-d3e867ef1f98.png)
![220b24-Printpreview-Mozilla Save to PDF](https://user-images.githubusercontent.com/80428201/220107972-9bdcdd77-9700-4724-abd5-8fdb450e5c76.png)

**[PDF24]**
extensions.printingtoolsng.printer.PDF24 = {"numCopies":1,"pageRanges":[1,1],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}
print.printer_PDF24.print_margin_* = 0.393750011920929 (inch)
print.printer_PDF24.print_paper_size_unit = 1 (mm)
print.printer_PDF24.print_paper_id = 9 (A4)
![220b24-PTNG-PDF24-margins](https://user-images.githubusercontent.com/80428201/220108142-2c2ca40f-2d56-4957-8b0a-a39722c5620c.png)
![220b24-Printpreview-PDF24](https://user-images.githubusercontent.com/80428201/220108168-4d0c2e77-6181-4ea5-8974-b7d30a7e98ea.png)

**[CubePDF]**
extensions.printingtoolsng.printer.CubePDF = {"numCopies":1,"pageRanges":[1,1],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}
print.printer_CubePDF.print_margin_* = 0.393750011920929 (inch)
print.printer_CubePDF.print_paper_size_unit = 1 (mm)
print.printer_CubePDF.print_paper_id = 9 (A4)
![220b24-PTNG-CubePDF-margins](https://user-images.githubusercontent.com/80428201/220108321-12f6a1a9-3359-4886-a6cc-aa8c0af00ecb.png)
![220b24-Printpreview-CubePDF](https://user-images.githubusercontent.com/80428201/220108374-17e36254-8cdc-472e-a669-6079cc4d8f88.png)

Margins set value:
0.393750011920929 (inch) ≠ 10.00 (mm) / e.g. 1.0001250302791597768cm

#166 mentions.

@kiki-ja

**Comment by cleidigh** [comment]
@kiki-ja 
Clarification 
Your last comment, is maybe misstype
10mm does equal  0.393750011920929 
@cleidigh

**Comment by kiki-ja** [comment]
@cleidigh

>Your last comment, is maybe misstype
>10mm does equal 0.393750011920929

You are right, it was a typo.
It was =, not ≠.

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

PTNG 2.2.0-**b25**
Thunderbird 102.8.0 64bit + Default Theme (System theme auto)
Windows 10 Pro 22H2 64bit

>Also if copies acting correctly with new delays?

The following are my test results.
1. Tested with MS_Publisher_Color_Printer
2. Set Copies to 2 in PTNG
3. Copies was unstable, going to 1 or 2 every time I previewed the print.

[PTNG set values / about:config]
extensions.printingtoolsng.printer.MS_Publisher_Color_Printer = {"numCopies":2,"pageRanges":[1,1],"headerStrLeft":"","headerStrCenter":"","headerStrRight":"","footerStrLeft":"","footerStrCenter":"","footerStrRight":""}

See screenshots.
![Copies-MS_Publisher_Color_Printer-PTNG-2](https://user-images.githubusercontent.com/80428201/220953117-4213092c-9611-415e-af00-3355c68464c9.png)

![Copies-MS_Publisher_Color_Printer-diff](https://user-images.githubusercontent.com/80428201/220953161-1a09165f-ac91-4374-8f0f-7288d81df527.png)

I do not know if this is an MS_Publisher_Color_Printer specific issue or not.
I expect it to work fine on other printers.

@kiki-ja

**Comment by kiki-ja** [comment]
@cleidigh

>This is why I made my recommendation to remove copies in #166

Unfortunately, it seems unstable, so I agree with the removal of the 'Copies' in the PTNG.

@kiki-ja
